src/ImageLoaders/IHex32/IntelHex32Parser.cs

```diff
@@ -0,0 +1,118 @@
+        /// </summary>
+        /// <param name="hexRecord">The hexadecimal record as a string.</param>
+        /// <param name="lineNum">(Optional) The line number in the IHex32 binary stream.</param>
+        /// <returns>
+        /// An <see cref="IntelHex32Record"/> .
+        /// </returns>
+        /// <exception cref="IntelHex32Exception">Thrown whenever an error is found in the IHex32 record.</exception>
+        public static IntelHex32Record ParseRecord(string hexRecord, int lineNum = 0)
+        {
+
+            List<byte> ParseHexData()
+            {
+                try
+                {
+                    return Enumerable.Range(1, hexRecord.Length - 1)
+                        .Where(i => (i & 1) != 0)
+                        .Select(i => Convert.ToByte(hexRecord.Substring(i, 2), 16))
+                        .ToList();
+                }
+                catch (Exception ex)
+                {
+                    throw new IntelHex32Exception($"Unable to parse hexedecimal numbers in Intel Hex line [{hexRecord}]", ex, lineNum);
+                }
+            }
+
+            if (hexRecord == null)
+                throw new IntelHex32Exception("Intel Hex line can not be null.", lineNum);
+            if (hexRecord.Length < MinHexRecordSize)
+                throw new IntelHex32Exception($"Intel Hex line [{hexRecord}] is less than {MinHexRecordSize} characters long.", lineNum);
+            if (hexRecord.Length % 2 == 0)
+                throw new IntelHex32Exception($"Intel Hex line has an even number of characters [{hexRecord}].", lineNum);
+            if (!hexRecord.StartsWith(":"))
+                throw new IntelHex32Exception($"Illegal Intel Hex line start character [{hexRecord}].", lineNum);
+
+            var hexData = ParseHexData();
+
+            if (hexData.Count != hexData[0] + 5)
+                throw new IntelHex32Exception($"Intel Hex line [{hexRecord}] does not have required record length of [{hexData[0] + 5}].", lineNum);
+
+            if (!Enum.IsDefined(typeof(IntelHex32RecordType), hexData[3]))
+                throw new IntelHex32Exception($"Intel Hex line has an invalid/unsupported record type value: [{hexData[3]}].", lineNum);
+
+            var rdr = new ImageReader(hexData.ToArray());
+            var datasize = rdr.ReadByte();
+
+            var newRecord = new IntelHex32Record
+            {
+                ByteCount = datasize,
+                Address = rdr.ReadBeUInt16(),
+                RecordType = (IntelHex32RecordType)rdr.ReadByte(),
+                Data = rdr.ReadBytes(datasize).ToList(),
+                CheckSum = rdr.ReadByte()
+            };
+
+            rdr.Offset = 0;
+            byte chk = 0;
+            while (rdr.IsValid)
+            {
+                chk += rdr.ReadByte();
+            }
+            if (chk != 0)
+                throw new IntelHex32Exception($"Checksum for Intel Hex line [{hexRecord}] is incorrect.", lineNum);
```

Comment:
Checksum verification was in separate method `VerifyChecksum`. Wasn't it better than putting all in one method?

Reply:
It's solely used by this record parser method and is simpler.
Personally I do prefer the new implementation as it is more conforming to what the Intel Hex specs implies: "the sum of all the bytes in the record should be 0". See page 5, last paragraph, of attache document.

[Intel HEX Standard.pdf](https://github.com/uxmal/reko/files/1754544/Intel.HEX.Standard.pdf)

Reply:
> Personally I do prefer the new implementation

Maybe this new implementation could be moved to separate method `VerifyCheckSum(byte [] bytes)`?